File: algorithms/dotnet/src/SortingAlgorithms/src/Algorithms.SortingAlgorithms/Quicksort.cs
namespace Algorithms.SortingAlgorithms;
public class QuickSort
{
    private static int Partition(int[] arr, int low, int high)
    {
        var pivot = arr[high];
        var i = low - 1;

        for (var j = low; j <= high - 1; j++)
        {
            if (arr[j] < pivot)
            {
                i++;
                (arr[i], arr[j]) = (arr[j], arr[i]);
            }
        }

        (arr[i + 1], arr[high]) = (arr[high], arr[i + 1]);
        return i + 1;
    }
    public static void Sort(int[] arr, int low, int high)
    {
        if (low < high)
        {
            var part = Partition(arr, low, high);
            Sort(arr, low, part - 1);
            Sort(arr, part + 1, high);
        }
    }
}

File: algorithms/java/src/main/java/org/sorting_algorithms/quicksort/Quicksort.java
package org.sorting_algorithms.quicksort;

/**
 * @author Dmytro Honchar <dmytro.honchar972@gmail.com>
 * Date: 10/1/2022
 */
public class Quicksort {
    private static int partition(int[] arr, int low, int high) {
        var pivot = arr[high];
        var i = low - 1;

        for (var j = low; j <= high - 1; j++) {
            if (arr[j] < pivot) {
                i++;
                var temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        var temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return i + 1;
    }

    public static void sort(int[] arr, int low, int high) {
        if (low < high) {
            var part = partition(arr, low, high);
            sort(arr, low, part - 1);
            sort(arr, part + 1, high);
        }
    }

    public static void main(String[] args) {
        int[] arr = {10, 7, 8, 9, 1, 5};
        sort(arr, 0, arr.length - 1);
        for (var i : arr)
            System.out.print(i + " ");
    }
}
